lightwood/encoders/time_series/transformer.py

```diff
@@ -0,0 +1,196 @@
+        loss = torch.dot((1.0 / lengths.float()), losses) / len(losses)
+        return loss
+
+    def _prepare_raw_data(self, data):
+        # Convert to array and determine max length:
+        # priming_data is a list of len B with lists of length in [0, T]
+        # If the length of the series varies a lot, it would be worth it to order them and apply a few optimisations
+        # we do not do so now
+        out_data = [
+            torch.tensor(e, dtype=torch.float, device=self.device) for e in data
+        ]
+        lengths = torch.tensor(
+            [len(e) for e in data], dtype=torch.long, device=self.device
+        )
+        return out_data, lengths
+
+    def prepare_encoder(
+        self, priming_data, feedback_hoop_function=None, batch_size=256
+    ):
+        """
+        The usual, run this on the initial training data for the encoder
+        :param priming_data: a list of 1-dimensional time series [[list1], ..., [listT]]
+        :param feedback_hoop_function: [if you want to get feedback on the training process]
+        :param batch_size
+        :return:
+        """
+        if self._prepared:
+            raise RuntimeError(
+                'You can only call "prepare_encoder" once for a given encoder.'
+            )
+
+        # We append the eos symbol
+        self._append_eos(priming_data)
+        priming_data, lengths_data = self._prepare_raw_data(priming_data)
+        npoints = len(priming_data)
+
+        # We pad the sequence with the function from PyTorch
+        self._encoder.train()
+        for epoch in range(self._train_iters):
+            total_loss = 0.0
+            for start_batch in range(0, npoints, batch_size):
+                self._optimizer.zero_grad()
+
+                train_batch = self._get_batch(priming_data, start_batch, batch_size)
+                len_batch = self._get_batch(lengths_data, start_batch, batch_size)
+                train_batch = pad_sequence(train_batch)
+
+                # TBPTT
+                losses = []
+                for start_chunk in range(0, train_batch.size(0) - 1, self.bptt):
+                    data, targets, lengths_chunk = self._get_chunk(
+                        train_batch, len_batch, start_chunk, self.bptt
+                    )
+                    data = data.unsqueeze(-1)
+                    output = self._encoder(data, lengths_chunk)
+                    output = output.squeeze(-1)
+                    losses.append(
+                        self._masked_criterion(output, targets, lengths_chunk)
+                    )
+                loss = sum(losses)
+                loss.backward()
+                torch.nn.utils.clip_grad_norm_(self._encoder.parameters(), 0.5)
+                self._optimizer.step()
+
+                total_loss += loss.item() * len(train_batch)
+
+            average_loss = total_loss / npoints
+            if average_loss < self._stop_on_error:
+                break
```

Comment:
Would you agree that a better approach here is:

1. Have a small validation subset
2. Look at the accuracy on that validation subset
3. If that accuracy drops for a given number of epochs, stop training and take the model that had the best accuracy on the validation subset
4. Or stop training if a given time limit is reached (rather than nr of epochs) and, again, pick the best model on the valdiation data
5. Run another epoch or two where the model is also trained on the validation  data

That is the approach we currently use for our mixer, I'm curios if you think it's preferable ? If so, no worries, no need to adopt it right now, was working on a generic way of implementing this for all encoders in `gym.py ` which I have to revisit at some poinrt.

Reply:
The approach I implemented is that I saw in `rnn_encoder.py`. I believe that it is a fairly rough one, but it is not clear to me how large are usually the sets sent to this `prepare` method. If they are very very small, then maybe it would be enough as it is, although the model would probably heavily overfit the data for a sufficiently large number of steps.

I think that, in genera, it would be better the approach that you have sketched above, i.e. that of a training with early stopping, to avoid overfitting to the training set when the datasets are quite small

Reply:
The `prepare` method is currently given the full training dataset (although not the validation dataset that is used for the "mixer", i.e. the component that comes after the encoders), so currently we design the encoders to work with the full dataset, it would be fairly trivial to also give them a validation datasets.

At any rate, the approach here is good, I was just curious what you thought about the above since (hopefully) it's some version of that which I'll be moving all trainable encoders too soon(tm).